### PR TITLE
feat: [DHIS2-14465] Category selection for the program

### PR DESCRIPTION
**Pull request by jasminenguyennn**
###  [DHIS2-14465](https://dhis2.atlassian.net/browse/DHIS2-14465): Category selection for the program is shown in:

1. Report new event
2. Schedule new event
3. View event
4. Edit event

The feature includes update, view mandatory category combo, show error message 

[DHIS2-14465]: https://dhis2.atlassian.net/browse/DHIS2-14465?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/core_modules/capture-core/components/DataEntry/withSaveHandler/withSaveHandler.js [hunk 1]
```diff
@@ -224,7 +224,6 @@ const getSaveHandler = (
 
         render() {
             const {
-                itemId,
```
Comment (jasminenguyennn) [inline src/core_modules/capture-core/components/DataEntry/withSaveHandler/withSaveHandler.js:227]: I want to expose `itemId` in the HOC so remove it here

**Review by simonadomnisoru — COMMENTED** [review]
Hey,
Great work on this PR 👏 ! I have a few comments, hope you can have a look at them. 

Firstly, it looks like the categories are not visible the first time when going to the `Enrollment: New Event` page. If the page is refreshed then the categories will be visible. The steps I took are in the video. Can you have a look?

Besides this, I made some small comments mainly related to changing the `Object` type to a strong type. The suggestions passed the check on my local host, but I'm not sure if they covered all the scenarios you had in mind. 

https://user-images.githubusercontent.com/6817073/222108277-d32c4407-f3f8-40e0-acaf-425fb06f3011.mov

Thank you!

File: src/core_modules/capture-core/components/FormFields/New/CategoryOptions/CategoryOptions.component.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+// @flow
+import * as React from 'react';
+import { withStyles } from '@material-ui/core/styles';
+import IconButton from '@material-ui/core/IconButton';
+import ClearIcon from '@material-ui/icons/Clear';
```
[the remark this answers is not in the document]
Comment (jasminenguyennn) [reply src/core_modules/capture-core/components/FormFields/New/CategoryOptions/CategoryOptions.component.js:5]: I'd keep the same icons from the [original design](https://github.com/dhis2/capture-app/blob/master/src/core_modules/capture-core/components/ScopeSelector/QuickSelector/Program/ProgramSelector.component.js#L284) and we will migrate the icons all together

File: src/core_modules/capture-core/components/DataEntries/Enrollment/EnrollmentDataEntry.component.js [hunk 3]
```diff
@@ -251,27 +289,49 @@ const getGeometrySettings = () => ({
...
+    },
+};
+
+type State = {
+    dataEntrySections: { [string]: {name: string, placement: $Values<typeof placements>}},
+}
+class FinalEnrollmentDataEntry extends React.Component<FinalTeiDataEntryProps, State> {
+    constructor(props) {
+        super(props);
+        if (props.programCategory) {
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/DataEntries/Enrollment/EnrollmentDataEntry.component.js:313]: NIT: Avoid modifying the `dataEntrySectionDefinitions` object and create a new object instead. Something like this.

```
const dataEntrySections = props.programCategory
    ? {
        ...dataEntrySectionDefinitions,
        [sectionKeysForEnrollmentDataEntry.CATEGORYCOMBO]: {
          placement: placements.TOP,
          name: props.programCategory.displayName,
        },
    }
    : dataEntrySectionDefinitions;

this.state = {
    dataEntrySections,
};
```

**Review by simonadomnisoru — APPROVED** [review]

File: src/core_modules/capture-core/components/DataEntryDhis2Helpers/withDataEntryFieldsIfApplicable.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+                <Component
+                    {...this.props}
+                />
+            );
+        }
+    };
+
+export const withDataEntryFieldsIfApplicable = (settings: Settings) =>
+    (InnerComponent: React.ComponentType<any>) =>
+        getDataEntryFieldsIfApplicable(settings, InnerComponent);
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/DataEntryDhis2Helpers/withDataEntryFieldsIfApplicable.js:39]: Don't think we really need this file (and probably at a later point we should remove the withDataEntryFieldIfApplicable too, but that is out of scope for this ticket). If we don't return anything from getFieldIds (see other comment) then we will not render any additional fields.

File: src/core_modules/capture-core/components/FormFields/New/CategoryOptions/CategoryOptions.component.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
+// @flow
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/FormFields/New/CategoryOptions/CategoryOptions.component.js:1]: If the category components for Scheduling have to be different than the other ones, let's put these files in the Scheduling folder.

File: src/core_modules/capture-core/components/WidgetEnrollmentEventNew/DataEntry/DataEntry.component.js [hunk 6]
```diff
@@ -298,12 +306,55 @@ const buildAssigneeSettingsFn = () => {
...
+                    ),
+                ),
+            ),
+        );
+    const categoryOptionsSettings = {
+        getComponent: () => categoryOptionsComponent,
+        getComponentProps: (props: Object, fieldId: string) => createComponentProps(props, {
+            ...props.categories?.find(category => category.id === fieldId) ?? {},
+            onSetFocus: () => {},
+            onRemoveFocus: () => {},
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/WidgetEnrollmentEventNew/DataEntry/DataEntry.component.js:333]: Let's implement these the same way as the other dataEntry fields. I think using the `withFocusSaver` should do the trick (you will see the blue background on focus when this is properly implemented)

**Review by JoakimSM — COMMENTED** [review]
Getting there, please ask if anything is unclear in my comments.

**Comment by JoakimSM** [comment]
1. If we try to save without touching the category field(s) the error is not shown (testing in the new event form). Can you try to replicate the behaviour from occuredAt?
2. NIT: Let's remove the 'before saving the event' part of the error message
3. When editing an event, the AOC selection is shown at the bottom of the page. We should display it beneath the basic info.
4. When scheduling, you can still click save/schedule without filling in the categories (you get an error from the api in the end). The user should be stopped here, the same way as we do in report new event. After having clicked "Schedule" all errors in the form should be visible (again, same way as for report new event).
5. NIT: When saving the event `attributeOptionCombo:""` is sent to the api. Can we remove this property altogether if it is not needed?
6. [comment](https://github.com/dhis2/capture-app/pull/3190#discussion_r1171429974). We are going to merge this before DHIS2-9661, can you comment out the logic and always hide the AOC (and we can re-enable in 9661)

**Comment by jasminenguyennn** [comment]
@JoakimSM 
> NIT: When saving the event attributeOptionCombo:"" is sent to the api. Can we remove this property altogether if it is not needed?
This field is needed when sending to api, removing it will result in failed request.

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2.41,2.40.0,2.40.1 versions

**Pull request merged**